airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/typing_deduping/BigQuerySqlGenerator.java

```diff
@@ -375,52 +366,51 @@ String dedupFinalTable(final StreamId id,
-                  FROM ${raw_table_id}
-                  WHERE
-                    JSON_QUERY(`_airbyte_data`, '$._ab_cdc_deleted_at') IS NOT NULL
-                    AND JSON_TYPE(JSON_QUERY(`_airbyte_data`, '$._ab_cdc_deleted_at')) != 'null'
-                    AND ${cursor_name} < ${cursor_cast}
-               );""");
+              )
+            ;""");
   }
 
   @VisibleForTesting
-  String dedupRawTable(final StreamId id, final String finalSuffix, LinkedHashMap<ColumnId, AirbyteType> streamColumns) {
-    /*
-     * Note that we need to keep the deletion raw records because of how async syncs work. Consider this
-     * sequence of source events: 1. Insert record id=1 2. Update record id=1 3. Delete record id=1
-     *
-     * It's possible for the destination to receive them out of order, e.g.: 1. Insert 2. Delete 3.
-     * Update
-     *
-     * We can generally resolve this using the cursor column (e.g. multiple updates in the wrong order).
-     * However, deletions are special because we propagate them as hard deletes to the final table. As a
-     * result, we need to keep the deletion in the raw table, so that a late-arriving update doesn't
-     * incorrectly reinsert the final record.
-     */
-    String cdcDeletedAtClause;
-    if (streamColumns.containsKey(CDC_DELETED_AT_COLUMN)) {
-      cdcDeletedAtClause = """
-                           AND (
-                             JSON_QUERY(`_airbyte_data`, '$._ab_cdc_deleted_at') IS NULL
-                             OR JSON_TYPE(JSON_QUERY(`_airbyte_data`, '$._ab_cdc_deleted_at')) = 'null'
-                           )
-                           """;
-    } else {
-      cdcDeletedAtClause = "";
+  String cdcDeletes(final StreamConfig stream,
+      final String finalSuffix,
+      final LinkedHashMap<ColumnId, AirbyteType> streamColumns) {
+
+    if (stream.destinationSyncMode() != DestinationSyncMode.APPEND_DEDUP){
+      return "";
     }
 
+    if (!streamColumns.containsKey(CDC_DELETED_AT_COLUMN)){
+      return "";
+    }
+
+    // we want to grab IDs for deletion from the raw table (not the final table itself) to hand out-of-order record insertions after the delete has been registered
+    return new StringSubstitutor(Map.of(
+        "final_table_id", stream.id().finalTableId(finalSuffix, QUOTE),
+        "raw_table_id", stream.id().rawTableId(QUOTE),
+        "quoted_cdc_delete_column", QUOTE + "_ab_cdc_deleted_at" + QUOTE)
+    ).replace(
+        // TODO replace `id`, `$.id` with PK
+        // TODO replace `INT64` with PK's type
+        """
+        DELETE FROM ${final_table_id} 
+        WHERE 
+          `id` IN (
+            SELECT
+              SAFE_CAST(JSON_VALUE(`_airbyte_data`, '$.id') as INT64) as id
+            FROM  ${raw_table_id}
+            WHERE
+              JSON_VALUE(`_airbyte_data`, '$._ab_cdc_deleted_at') IS NOT NULL
+              OR JSON_TYPE(JSON_QUERY(`_airbyte_data`, '$._ab_cdc_deleted_at')) = 'null'
+          )
+        ;"""
```

Comment:
I need help filling in these TODOs

Reply:
:eyes: I'll take a look at this next week. PK handling is a bit sketchy - the DATs branch had to make some fixes in this area https://github.com/airbytehq/airbyte/pull/27852/files#diff-114b383c1f870c36a452c850f27df1ef187af0ba843ab204e0af0ceb6e51b7a9R392